### DISCUSSION TITLE
upload objects not always success and listing object also

### DISCUSSION THREAD

**Discussion by mazataza** [minio/minio#12496] [Q&A] [answered]
we are effected by minio:RELEASE.2021-05-27T22-06-31Z (i created the issue for the currupted files)
from that release and after new release we still have failure behaviour.

1- during uploading files sometime (not always reproducable) we get network error
  this happend to different users from different region. (so it is not related to one user.
 what i see on logs is following:
> API: SYSTEM()
Jun 12 19:55:48 minio04 minio[991]: Time: 19:55:48 UTC 06/12/2021
Jun 12 19:55:48 minio04 minio[991]: DeploymentID: 7c507082-fe65-439a-9391-3b44b1b1d2ed
Jun 12 19:55:48 minio04 minio[991]: Error: Disk: http://minio03:9000/drive2  -> alwaqiyah/Aqsa_Call_Okab11052021SD.mp4/530bdc4>
Jun 12 19:55:48 minio04 minio[991]:        2: cmd/bitrot-streaming.go:179:cmd.(*streamingBitrotReader).ReadAt()
Jun 12 19:55:48 minio04 minio[991]:        1: cmd/erasure-decode.go:163:cmd.(*parallelReader).Read.func1()

thus files are those which was currupted by the effected minio version and we still not recovered them.
we don't know now which are exactly are currupted as all files we see them on minio browser or tring to copy them.

> mc cp s3min/alwaqiyah/Aqsa_Call_Okab11052021SD.mp4 etc/Aqsa_Call_Okab11052021SD.mp4
...kab11052021SD.mp4:  0 B / 13.75 MiB
c: <ERROR> Failed to copy `http://192.168.10.1:9000/alwaqiyah/Aqsa_Call_Okab11052021SD.mp4`. Resource requested is unreadable, please reduce your request rate

2-
`
mc ls alias/bucketname 
`
take very long to anwser and sometime never return and i should kill it with "ctrl +c"

i write a small javascript code to list object

>var Minio = require('minio')
var minioClient = new Minio.Client({
    endPoint: '192.168.10.1',
    port: 9000,
    useSSL: false,
    accessKey: '??',
    secretKey: '??'
});
var stream = minioClient.listObjects("alwaqiyah", "", false);
var count = 0;
stream.on('data',  (obj) => {
    count++;
    console.log(obj);
})
stream.on('error',  (err) => { 
    console.log("count: " + count)
    console.log(err) 
})
stream.on('end',  () => {
    console.log("number of items : " + count);
}
)
sometimes the listing is complete and sometimes i got such errors

> count: 1000
Error: read ECONNRESET
    at TCP.onStreamRead (internal/stream_base_commons.js:209:20) {
  errno: -4077,
  code: 'ECONNRESET',
  syscall: 'read'
}

actually we have more than 6000 objects in that bucket.

this behaviour is also happend from running this script from inside minio server it self.

---
we run minio as linux service (not docker).

what i try to do is to write a scipt which check files by uploading them and if not uploaded we mark it to be upload again by content teams.

I also not opened a ticket because it may be not related to minio itself, but I hope for help here.

**Comment by harshavardhana** [comment, marked as answer]
> we are effected by minio:RELEASE.2021-05-27T22-06-31Z (i created the issue for the currupted files)
from that release and after new release we still have failure behaviour.

As I explained all the objects that have been healed are unrecoverable it will not happen in the future for your existing objects that were not healed, but all the healed objects would have been corrupted from the legacy version of the objects.

There is not an automatic way to fix it, it is unfortunate because you did not validate this elsewhere in a local setup before rolling it out in production.  You just have to purge all those content and re-upload all of them from backups (if you have). If there are no backups those objects should be deleted anyways as they will not be readable anyways.

**Reply by mazataza** [reply]
That we want to do. We want to know those objects to upload them again

**Comment by harshavardhana** [comment]
This is why we suggest all our users buy our subscription for peace of mind - so that we can help and guide you during upgrades to avoid production downtimes - relying on an open-source project without getting guidance on the setup and deployment from the MinIO Engineers is a poor way to manage a production infrastructure.

I recommend that you get subscription for your future needs such that you can avoid this kind of situation in future https://min.io/pricing

**Reply by mazataza** [reply]
I mean with failure behaviour here not healing issue but during upload Ing New object we got network reset. This happend when using the software s3 Browser or s3 aws php api

**Reply by mazataza** [reply]
> This is why we suggest all our users buy our subscription for peace of mind - so that we can help and guide you during upgrades to avoid production downtimes - relying on an open-source project without getting guidance on the setup and deployment from the MinIO Engineers is a poor way to manage a production infrastructure.
> 
> I recommend that you get subscription for your future needs such that you can avoid this kind of situation in future https://min.io/pricing

We have 7 Terabyte and we cant buy 100 Dollar per month. We Hosting public sites and none commerical sites

**Reply by harshavardhana** [reply]
> I mean with failure behaviour here not healing issue but during upload Ing New object we got network reset. This happend when using the software s3 Browser or s3 aws php api

You will get network reset if one of the server is going down, as a HTTP client it's your responsibility to retry the operation again.

In-flight requests are always disconnected if you do rolling restarts upon upgrades etc. This is common and should be dealt with a proper client retry logic.

**Reply by harshavardhana** [reply]
> 
> We have 7 Terabyte and we cant buy 100 Dollar per month. We Hosting public sites and none commerical sites

Understood thought I suggest it..

**Reply by mazataza** [reply]
> 
> 
> > I mean with failure behaviour here not healing issue but during upload Ing New object we got network reset. This happend when using the software s3 Browser or s3 aws php api
> 
> You will get network reset if one of the server is going down, as a HTTP client it's your responsibility to retry the operation again.
> 
> In-flight requests are always disconnected if you do rolling restarts upon upgrades etc. This is common and should be dealt with a proper client retry logic.

our minio server runs since last update on 08.06.2021 without a restart, during last days.
thus i don't think it is related to server goes down.
our network from public is as following

nginx (reverse proxy) --> haproxy -> minio servers

**Reply by harshavardhana** [reply]
Then you need to look at nginx timeouts, haproxy timeouts etc.

**Comment by mazataza** [comment]
what not answered is why soemtimes i got network reset when runing the javascript code above and set endpoint directly to a minio instance (not nginx or haproxy).

I hope someone has a tip about the reason. 
inside minio servers we do not run any other network services.

**Comment by mazataza** [comment]
this is a nginx configuration i tried 

> server {
        listen 80;
        server_name xxxx;
        access_log /var/log/nginx/reverse-access.log;
        error_log /var/log/nginx/reverse-error.log;
       # To allow special characters in headers
       ignore_invalid_headers off;
       # Allow any size file to be uploaded.
       # Set to a value such as 1000m; to restrict file size to a specific value
       client_max_body_size 5G;
       # To disable buffering
       proxy_buffering off;
        location / {
                    proxy_set_header X-Real-IP $remote_addr;
                    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
                    proxy_set_header X-Forwarded-Proto $scheme;
                    proxy_set_header Host s3.htcmo.com;
                    #proxy_connect_timeout 300;
                    proxy_read_timeout 3600s;
                    proxy_connect_timeout 75s;
                    # Default is HTTP/1, keepalive is only enabled in HTTP/1.1
                    proxy_http_version 1.1;
                    proxy_set_header Connection "";
                    chunked_transfer_encoding off;
                    proxy_pass http://minio01:9000;
        }
}

when uploading from minio browser upload stuck on 100% and i see in browser console 504 gateway rime error

[HTTP/1.1 504 Gateway Time-out 172278ms
this happend for a file with ca 400mb

for me it looks like minio instances are no more stable in our servers after the curruption issue  appeared and also still the server is uninstable after upgrading to fixed minio version.
for whatever bucket we upload to them.

any advice?

**Reply by harshavardhana** [reply]
They are definitely stable, what you need is to increase the timeout from nginx..

You have a top level timeout.

**Comment by mazataza** [comment]
currently since yesterday and after restart minio with this release RELEASE.2021-06-09T18-51-39Z
before it was RELEASE.2021-06-07T21-40-51Z.
we can upload big files without issue. yes between 100% and complete upload it can take time (2.5G file) but it complete at the end.

is there any noticable changes in network behaviour of new release ?

but i think the network link between minio servers and the internet got slowance from time to time and we will also contact the datacenter company where we host our dedicated servers.

**Reply by harshavardhana** [reply]
> we can upload big files without issue. yes between 100% and complete upload it can take time (2.5G file) but it complete at the end.

I don't think we made any change @mazataza 

> is there any noticable changes in network behaviour of new release ?

Nope - most probably you experienced a crash that was fixed in RELEASE.2021-06-09T18-51-39Z

**Comment by mazataza** [comment]
it seams we had ddos attack on minio servers. we are now configuring firewall and allow connections only from our nginx or haproxy.
